File: lld/week-02-03-patterns/prototype/java/src/PrototypeTest.java
import java.util.List;
import java.util.Map;

/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out PrototypeTest` directly.
 */
public class PrototypeTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testCloneProducesEqualButDistinctValues();
        testMutatingCloneTagsDoesNotAffectOriginal();
        testMutatingCloneSectionsDoesNotAffectOriginal();
        testMutatingClonePropsDoesNotAffectOriginal();
        testMutatingCloneAuthorDoesNotAffectOriginal();
        System.out.println("All PrototypeTest cases passed.");
    }

    private static Document newTestDocument() {
        return new Document(
                "Q3 Report",
                new Metadata("Vinay", List.of("finance", "draft")),
                List.of("Intro", "Numbers"),
                Map.of("status", "draft"));
    }

    private static void testCloneProducesEqualButDistinctValues() {
        Document original = newTestDocument();
        Document clone = original.deepClone();

        assertEquals(original.getTitle(), clone.getTitle(), "title");
        assertEquals(original.getMeta().getAuthor(), clone.getMeta().getAuthor(), "author");
        if (clone == original) {
            throw new AssertionError("clone should be a distinct object from original");
        }
        if (clone.getMeta() == original.getMeta()) {
            throw new AssertionError("clone.meta should be a distinct object from original.meta");
        }
    }

    private static void testMutatingCloneTagsDoesNotAffectOriginal() {
        Document original = newTestDocument();
        Document clone = original.deepClone();

        clone.getMeta().getTags().set(0, "mutated");
        clone.getMeta().getTags().add("extra");

        assertEquals("finance", original.getMeta().getTags().get(0), "original tag 0 unchanged");
        assertEquals(2, original.getMeta().getTags().size(), "original tags size unchanged");
    }

    private static void testMutatingCloneSectionsDoesNotAffectOriginal() {
        Document original = newTestDocument();
        Document clone = original.deepClone();

        clone.getSections().set(0, "Rewritten Intro");
        clone.getSections().add("Appendix");

        assertEquals("Intro", original.getSections().get(0), "original section 0 unchanged");
        assertEquals(2, original.getSections().size(), "original sections size unchanged");
    }

    private static void testMutatingClonePropsDoesNotAffectOriginal() {
        Document original = newTestDocument();
        Document clone = original.deepClone();

        clone.getProps().put("status", "final");
        clone.getProps().put("reviewer", "Asha");

        assertEquals("draft", original.getProps().get("status"), "original status unchanged");
        assertEquals(false, original.getProps().containsKey("reviewer"), "original should not gain new key");
    }

    private static void testMutatingCloneAuthorDoesNotAffectOriginal() {
        Document original = newTestDocument();
        Document clone = original.deepClone();

        clone.getMeta().setAuthor("Someone Else");

        assertEquals("Vinay", original.getMeta().getAuthor(), "original author unchanged");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        boolean equal = (expected == null) ? (actual == null) : expected.equals(actual);
        if (!equal) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}

File: lld/week-02-03-patterns/prototype/go/prototype_test.go
package prototype

import "testing"

func newTestDocument() *Document {
	return &Document{
		Title:    "Q3 Report",
		Meta:     &Metadata{Author: "Vinay", Tags: []string{"finance", "draft"}},
		Sections: []string{"Intro", "Numbers"},
		Props:    map[string]string{"status": "draft"},
	}
}

func TestCloneProducesEqualButDistinctValues(t *testing.T) {
	original := newTestDocument()
	clone := original.Clone()

	if clone.Title != original.Title {
		t.Fatalf("expected equal titles, got %s vs %s", clone.Title, original.Title)
	}
	if clone.Meta.Author != original.Meta.Author {
		t.Fatalf("expected equal authors")
	}
	if clone == original {
		t.Fatalf("clone should be a distinct pointer from original")
	}
	if clone.Meta == original.Meta {
		t.Fatalf("clone.Meta should be a distinct pointer from original.Meta")
	}
}

func TestMutatingCloneTagsDoesNotAffectOriginal(t *testing.T) {
	original := newTestDocument()
	clone := original.Clone()

	clone.Meta.Tags[0] = "mutated"
	clone.Meta.Tags = append(clone.Meta.Tags, "extra")

	if original.Meta.Tags[0] != "finance" {
		t.Fatalf("mutating clone's tags leaked into original: %v", original.Meta.Tags)
	}
	if len(original.Meta.Tags) != 2 {
		t.Fatalf("appending to clone's tags leaked into original: %v", original.Meta.Tags)
	}
}

func TestMutatingCloneSectionsDoesNotAffectOriginal(t *testing.T) {
	original := newTestDocument()
	clone := original.Clone()

	clone.Sections[0] = "Rewritten Intro"
	clone.Sections = append(clone.Sections, "Appendix")

	if original.Sections[0] != "Intro" {
		t.Fatalf("mutating clone's sections leaked into original: %v", original.Sections)
	}
	if len(original.Sections) != 2 {
		t.Fatalf("appending to clone's sections leaked into original: %v", original.Sections)
	}
}

func TestMutatingClonePropsDoesNotAffectOriginal(t *testing.T) {
	original := newTestDocument()
	clone := original.Clone()

	clone.Props["status"] = "final"
	clone.Props["reviewer"] = "Asha"

	if original.Props["status"] != "draft" {
		t.Fatalf("mutating clone's props leaked into original: %v", original.Props)
	}
	if _, ok := original.Props["reviewer"]; ok {
		t.Fatalf("adding a key to clone's props leaked into original: %v", original.Props)
	}
}

func TestMutatingCloneAuthorDoesNotAffectOriginal(t *testing.T) {
	original := newTestDocument()
	clone := original.Clone()

	clone.Meta.Author = "Someone Else"

	if original.Meta.Author != "Vinay" {
		t.Fatalf("mutating clone's author leaked into original: %s", original.Meta.Author)
	}
}
